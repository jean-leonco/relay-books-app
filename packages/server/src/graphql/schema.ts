import { GraphQLSchema } from 'graphql';

import applyMiddlewares from '../core/middlewares/applyMiddlewares';

// import SubscriptionType from './subscription/SubscriptionType';
import MutationType from './type/MutationType';
import QueryType from './type/QueryType';

const _schema = new GraphQLSchema({
  query: QueryType,
  mutation: MutationType,
  // subscription: SubscriptionType,
});

applyMiddlewares(_schema);

export const schema = _schema;
