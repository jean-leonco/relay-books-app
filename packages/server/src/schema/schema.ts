import { GraphQLSchema } from 'graphql';

import applyMiddlewares from '../graphql/middlewares/applyMiddlewares';

import MutationType from './MutationType';
import QueryType from './QueryType';

const schema = new GraphQLSchema({
  query: QueryType,
  mutation: MutationType,
});

applyMiddlewares(schema);

export default schema;
