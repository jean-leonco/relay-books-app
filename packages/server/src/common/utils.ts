import { Model, Types } from 'mongoose';
import { fromGlobalId } from 'graphql-relay';

// https://stackoverflow.com/a/2593661/710693
export const escapeRegex = (str: string) => `${str}`.replace(/[.?*+^$[\]\\(){}|-]/g, '\\$&');

export const PLATFORM = {
  APP: 'APP',
  UNKNOWN: 'UNKNOWN',
};

export const getPlatform = (platform?: string | null): string => {
  if (typeof platform !== 'string') {
    return PLATFORM.UNKNOWN;
  }

  //if (platform.includes('ios') || platform.includes('android')) {
  //  return PLATFORM.APP;
  //}

  if (platform === PLATFORM.APP) {
    return PLATFORM.APP;
  }

  return PLATFORM.UNKNOWN;
};

export function kilometersToEarthRadians(kilometers: number) {
  return kilometers / 6378.1;
}

// returns an ObjectId given an param of unknown type
export const getObjectId = (target: string | Model<any> | Types.ObjectId | undefined | null): Types.ObjectId | null => {
  if (!target) {
    return null;
  }

  if (target instanceof Types.ObjectId) {
    return new Types.ObjectId(target.toString());
  }

  if (typeof target === 'object') {
    return target && target._id ? new Types.ObjectId(target._id) : null;
  }

  if (typeof target === 'string') {
    const result = fromGlobalId(target);

    if (result.type && result.id && Types.ObjectId.isValid(result.id)) {
      return new Types.ObjectId(result.id);
    }

    if (Types.ObjectId.isValid(target)) {
      return new Types.ObjectId(target);
    }

    return null;
  }

  return null;
};

type EnumObject = {
  [key: string]: string;
};

type EnumObjectResult = {
  [key: string]: {
    value: string;
  };
};

export const enumBuilderValues = (constants: EnumObject): EnumObjectResult =>
  Object.keys(constants).reduce(
    (prev, curr) => ({
      ...prev,
      [curr]: {
        value: constants[curr],
      },
    }),
    {},
  );
