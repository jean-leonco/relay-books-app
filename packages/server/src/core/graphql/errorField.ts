import { GraphQLString } from 'graphql';

const errorField = {
  error: {
    type: GraphQLString,
    description: 'Default error field resolver for mutations',
    resolve: ({ error }) => error,
  },
};

export default errorField;
