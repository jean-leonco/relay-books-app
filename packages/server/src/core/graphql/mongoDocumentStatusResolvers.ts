import { GraphQLBoolean, GraphQLString } from 'graphql';

export const mongoDocumentStatusResolvers = {
  isActive: {
    type: GraphQLBoolean,
    description: 'The soft delete status.',
    resolve: (obj) => obj.isActive,
  },
  removedAt: {
    type: GraphQLString,
    description: 'The hard delete status.',
    resolve: (obj) => (obj.removedAt ? obj.removedAt.toISOString() : null),
  },
  createdAt: {
    type: GraphQLString,
    description: 'The date that the node was created.',
    resolve: (obj) => (obj.createdAt ? obj.createdAt.toISOString() : null),
  },
  updatedAt: {
    type: GraphQLString,
    description: 'The date that the node was last updated.',
    resolve: (obj) => (obj.updatedAt ? obj.updatedAt.toISOString() : null),
  },
};
