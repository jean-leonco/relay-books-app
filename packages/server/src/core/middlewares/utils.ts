import { GraphQLResolveInfo } from 'graphql';
import { ObjectSchema, Shape } from 'yup';

import { GraphQLContext } from '../../types';

interface GraphQLMutationConfig {
  validationSchema?(context: GraphQLContext): ObjectSchema<Shape<any, any>>;
  authenticatedOnly?: boolean;
}

export function getMutationFields(info: GraphQLResolveInfo, mutationName: string): GraphQLMutationConfig | null {
  const mutationType = info.schema.getMutationType();
  if (!mutationType) {
    return null;
  }

  const fields = mutationType.getFields();
  return fields[mutationName];
}
