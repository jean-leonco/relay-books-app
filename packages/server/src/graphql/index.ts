/* eslint-disable no-console */
import 'core-js/stable';
import { createServer } from 'http';

import { execute, subscribe } from 'graphql';
import { SubscriptionServer } from 'subscriptions-transport-ws';

import { GRAPHQL_PORT, GRAPHQL_HOST } from '../common/config';
import connectDatabase from '../common/database';

import app from './app';
import { schema } from './schema';

type ConnectionParams = {
  authorization?: string;
};

const runServer = async () => {
  try {
    console.log('\n🔗 Connecting to database...');
    await connectDatabase();
  } catch (error) {
    console.error('Could not connect to database', { error });
    throw error;
  }

  const server = createServer(app.callback());

  server.listen(GRAPHQL_PORT, () => {
    console.info(`\n🚀 Server started at http://${GRAPHQL_HOST}:${GRAPHQL_PORT}`);

    if (process.env.NODE_ENV !== 'production') {
      console.info(`\n🎠 GraphQL Playground available at http://${GRAPHQL_HOST}:${GRAPHQL_PORT}/playground\n`);
    }

    SubscriptionServer.create(
      {
        onConnect: async (connectionParams: ConnectionParams) => {
          console.info('Client subscription connected', connectionParams);
          // @TODO - get user from connectionParams?.authorization
          // @TODO - return context with user
        },
        onDisconnect: () => console.info('Client subscription disconnected'),
        execute,
        subscribe,
        schema,
      },
      {
        server,
        path: '/subscriptions',
      },
    );
  });
};

(async () => {
  console.log('\n📡 Server starting...');
  await runServer();
})();
