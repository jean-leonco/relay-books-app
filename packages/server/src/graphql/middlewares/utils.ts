import { GraphQLField, GraphQLResolveInfo } from 'graphql';

import { Extensions } from '../../types';

interface Field extends GraphQLField<any, any> {
  extensions: Extensions | null | undefined;
}

export function getMutationFields(info: GraphQLResolveInfo, mutationName: string): Field | null {
  const mutationType = info.schema.getMutationType();
  if (!mutationType) {
    return null;
  }

  const fields = mutationType.getFields();
  return fields[mutationName];
}
