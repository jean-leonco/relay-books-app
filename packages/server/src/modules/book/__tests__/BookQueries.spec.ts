import MockDate from 'mockdate';
import { graphql } from 'graphql';
import { toGlobalId } from 'graphql-relay';

import {
  sanitizeTestObject,
  connectMongoose,
  clearDbAndRestartCounters,
  disconnectMongoose,
  gql,
  resetRunningDate,
  bumpDate,
  N_DAYS_IN_MILLISECONDS,
} from '@workspace/test-utils';

import { createBook, createCategory, createReading, createUser, getContext } from '../../../test/utils';

import schema from '../../../schema/schema';

beforeAll(connectMongoose);

beforeEach(async () => {
  await clearDbAndRestartCounters();
  MockDate.reset();
  resetRunningDate();
  jest.clearAllMocks();
});

afterAll(disconnectMongoose);

describe('BookQueries', () => {
  it('should get book from node interface', async () => {
    const user = await createUser();
    const book = await createBook();

    const query = gql`
      query Q($id: ID!) {
        book: node(id: $id) {
          ... on Book {
            id
            name
            author
            description
          }
        }
      }
    `;

    const rootValue = {};
    const context = await getContext({ user });
    const variables = {
      id: toGlobalId('Book', book._id),
    };

    const result = await graphql(schema, query, rootValue, context, variables);

    expect(result.errors).toBeUndefined();
    expect(result.data?.book).not.toBe(null);
    expect(sanitizeTestObject(result.data)).toMatchSnapshot();
  });

  it('should not get book without user', async () => {
    const book = await createBook();

    const query = gql`
      query Q($id: ID!) {
        book: node(id: $id) {
          ... on Book {
            id
            name
            author
            description
          }
        }
      }
    `;

    const rootValue = {};
    const context = await getContext();
    const variables = {
      id: toGlobalId('Book', book._id),
    };

    const result = await graphql(schema, query, rootValue, context, variables);

    expect(result.errors).toBeUndefined();
    expect(result.data?.book).toBe(null);
  });

  it('should get null book if isActive is false', async () => {
    const user = await createUser();
    const event = await createBook({ isActive: false });

    const query = gql`
      query Q($id: ID!) {
        book: node(id: $id) {
          ... on Book {
            id
            name
            author
            description
          }
        }
      }
    `;

    const rootValue = {};
    const context = await getContext({ user });
    const variables = {
      id: toGlobalId('Book', event._id),
    };

    const result = await graphql(schema, query, rootValue, context, variables);

    expect(result.errors).toBeUndefined();
    expect(result.data?.book).toBe(null);
  });

  it('should query a connection of books', async () => {
    const user = await createUser();

    for (let i = 0; i < 6; i++) {
      await createBook();
    }

    const query = gql`
      query Q {
        books(first: 5) {
          count
          edges {
            node {
              id
              name
              author
              description
            }
          }
        }
      }
    `;

    const rootValue = {};
    const context = await getContext({ user });
    const variables = {};

    const result = await graphql(schema, query, rootValue, context, variables);

    expect(result.errors).toBeUndefined();
    expect(result.data?.books).not.toBe(null);
    expect(result.data?.books.edges.length).toBe(5);
    expect(sanitizeTestObject(result.data)).toMatchSnapshot();
  });

  it('should query a connection of books with search filter', async () => {
    const user = await createUser();

    await createBook({ name: 'TEST 1' });
    await createBook({ author: 'TEST 3' });
    await createBook({ description: 'TEST 2' });
    await createBook({ name: 'DO NOT SEARCH 1' });
    await createBook({ name: 'DO NOT SEARCH 2' });
    await createBook({ name: 'DO NOT SEARCH 3' });

    const query = gql`
      query Q($filters: BookFilters) {
        books(filters: $filters) {
          count
          edges {
            node {
              id
              name
              author
              description
            }
          }
        }
      }
    `;

    const rootValue = {};
    const context = await getContext({ user });
    const variables = {
      filters: { search: 'TEST' },
    };

    const result = await graphql(schema, query, rootValue, context, variables);

    expect(result.errors).toBeUndefined();
    expect(result.data?.books).not.toBe(null);
    expect(result.data?.books.edges.length).toBe(3);
    expect(sanitizeTestObject(result.data)).toMatchSnapshot();
  });

  it('should query a connection of books with orderBy filter', async () => {
    const user = await createUser();
    const book1 = await createBook();
    const book2 = await createBook();
    const book3 = await createBook();

    const query = gql`
      query Q($filters: BookFilters) {
        books(filters: $filters) {
          count
          edges {
            node {
              id
              name
              author
              description
            }
          }
        }
      }
    `;

    const rootValue = {};
    const context = await getContext({ user });
    const variables = {
      filters: {
        orderBy: { field: 'CREATED_AT', direction: 'ASC' },
      },
    };

    const result = await graphql(schema, query, rootValue, context, variables);

    expect(result.errors).toBeUndefined();
    expect(result.data?.books).not.toBe(null);
    expect(result.data?.books.edges.length).toBe(3);
    expect(result.data?.books.edges[0].node.name).toBe(book1.name);
    expect(result.data?.books.edges[1].node.name).toBe(book2.name);
    expect(result.data?.books.edges[2].node.name).toBe(book3.name);
    expect(sanitizeTestObject(result.data)).toMatchSnapshot();
  });

  it('should query a connection of books with category filter', async () => {
    const user = await createUser();

    const horrorCategory = await createCategory({ name: 'Horror' });
    for (let i = 0; i < 5; i++) {
      await createBook({ categoryId: horrorCategory._id });
    }

    const comedyCategory = await createCategory({ name: 'Comedy' });
    for (let i = 0; i < 10; i++) {
      await createBook({ categoryId: comedyCategory._id });
    }

    const query = gql`
      query Q($filters: BookFilters) {
        books(filters: $filters) {
          count
          edges {
            node {
              id
              name
              author
              description
            }
          }
        }
      }
    `;

    const rootValue = {};
    const context = await getContext({ user });
    const variables = {
      filters: {
        category: toGlobalId('Category', horrorCategory._id),
      },
    };

    const result = await graphql(schema, query, rootValue, context, variables);

    expect(result.errors).toBeUndefined();
    expect(result.data?.books).not.toBe(null);
    expect(result.data?.books.edges.length).toBe(5);
    expect(sanitizeTestObject(result.data)).toMatchSnapshot();
  });

  it('should query a connection of books with trending filter', async () => {
    const user = await createUser();

    const book1 = await createBook();
    for (let i = 0; i < 10; i++) {
      await createReading({ bookId: book1._id });
    }

    const book2 = await createBook();
    for (let i = 0; i < 5; i++) {
      await createReading({ bookId: book2._id });
    }

    const book3 = await createBook();
    for (let i = 0; i < 3; i++) {
      await createReading({ bookId: book3._id });
    }

    const query = gql`
      query Q($filters: BookFilters) {
        books(filters: $filters) {
          count
          edges {
            node {
              id
              name
            }
          }
        }
      }
    `;

    const rootValue = {};
    const context = await getContext({ user });
    const variables = {
      filters: {
        trending: true,
      },
    };

    const result = await graphql(schema, query, rootValue, context, variables);

    expect(result.errors).toBeUndefined();
    expect(result.data?.books).not.toBe(null);
    expect(result.data?.books.edges.length).toBe(3);
    expect(result.data?.books.edges[0].node.name).toBe(book1.name);
    expect(result.data?.books.edges[1].node.name).toBe(book2.name);
    expect(result.data?.books.edges[2].node.name).toBe(book3.name);
  });

  it('should query with trending filter and use creation date to sort if books has the same readings amount', async () => {
    const user = await createUser();

    const book1 = await createBook();
    for (let i = 0; i < 10; i++) {
      await createReading({ bookId: book1._id });
    }

    const book2 = await createBook();
    for (let i = 0; i < 5; i++) {
      await createReading({ bookId: book2._id });
    }

    const book3 = await createBook();
    for (let i = 0; i < 3; i++) {
      await createReading({ bookId: book3._id });
    }

    const book4 = await createBook();
    for (let i = 0; i < 10; i++) {
      await createReading({ bookId: book4._id });
    }

    const query = gql`
      query Q($filters: BookFilters) {
        books(filters: $filters) {
          count
          edges {
            node {
              id
              name
            }
          }
        }
      }
    `;

    const rootValue = {};
    const context = await getContext({ user });
    const variables = {
      filters: {
        trending: true,
      },
    };

    const result = await graphql(schema, query, rootValue, context, variables);

    expect(result.errors).toBeUndefined();
    expect(result.data?.books).not.toBe(null);
    expect(result.data?.books.edges.length).toBe(4);
    expect(result.data?.books.edges[0].node.name).toBe(book4.name);
    expect(result.data?.books.edges[1].node.name).toBe(book1.name);
    expect(result.data?.books.edges[2].node.name).toBe(book2.name);
    expect(result.data?.books.edges[3].node.name).toBe(book3.name);
  });

  it('should query with trending filter and consider only last 7 days', async () => {
    MockDate.set(new Date('10/03/2020'));

    const user = await createUser();

    const book1 = await createBook();
    for (let i = 0; i < 10; i++) {
      await createReading({ bookId: book1._id });
    }

    const book2 = await createBook();
    for (let i = 0; i < 5; i++) {
      await createReading({ bookId: book2._id });
    }

    const book3 = await createBook();
    for (let i = 0; i < 3; i++) {
      await createReading({ bookId: book3._id });
    }

    bumpDate(new Date(), N_DAYS_IN_MILLISECONDS(8));

    for (let i = 0; i < 2; i++) {
      await createReading({ bookId: book1._id });
    }

    for (let i = 0; i < 5; i++) {
      await createReading({ bookId: book2._id });
    }

    for (let i = 0; i < 15; i++) {
      await createReading({ bookId: book3._id });
    }

    const query = gql`
      query Q($filters: BookFilters) {
        books(filters: $filters) {
          count
          edges {
            node {
              id
              name
            }
          }
        }
      }
    `;

    const rootValue = {};
    const context = await getContext({ user });
    const variables = {
      filters: {
        trending: true,
      },
    };

    const result = await graphql(schema, query, rootValue, context, variables);

    expect(result.errors).toBeUndefined();
    expect(result.data?.books).not.toBe(null);
    expect(result.data?.books.edges.length).toBe(3);
    expect(result.data?.books.edges[0].node.name).toBe(book3.name);
    expect(result.data?.books.edges[1].node.name).toBe(book2.name);
    expect(result.data?.books.edges[2].node.name).toBe(book1.name);
  });

  it('should query only active books', async () => {
    const user = await createUser();

    for (let i = 0; i < 5; i++) {
      await createBook();
    }

    for (let i = 0; i < 10; i++) {
      await createBook({ isActive: false });
    }

    const query = gql`
      query Q {
        books {
          count
          edges {
            node {
              id
              name
              author
              description
            }
          }
        }
      }
    `;

    const rootValue = {};
    const context = await getContext({ user });
    const variables = {};

    const result = await graphql(schema, query, rootValue, context, variables);

    expect(result.errors).toBeUndefined();
    expect(result.data?.books).not.toBe(null);
    expect(result.data?.books.edges.length).toBe(5);
  });
});
