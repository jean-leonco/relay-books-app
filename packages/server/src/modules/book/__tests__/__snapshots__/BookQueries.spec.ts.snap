// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Book queries should get book from node interface 1`] = `
Object {
  "book": Object {
    "author": "author 1",
    "description": "description 1",
    "id": "FROZEN-ID",
    "name": "name 1",
  },
}
`;

exports[`Book queries should query a connection of books 1`] = `
Object {
  "books": Object {
    "count": 6,
    "edges": Array [
      Object {
        "node": Object {
          "author": "author 6",
          "description": "description 6",
          "id": "FROZEN-ID",
          "name": "name 6",
        },
      },
      Object {
        "node": Object {
          "author": "author 5",
          "description": "description 5",
          "id": "FROZEN-ID",
          "name": "name 5",
        },
      },
      Object {
        "node": Object {
          "author": "author 4",
          "description": "description 4",
          "id": "FROZEN-ID",
          "name": "name 4",
        },
      },
      Object {
        "node": Object {
          "author": "author 3",
          "description": "description 3",
          "id": "FROZEN-ID",
          "name": "name 3",
        },
      },
      Object {
        "node": Object {
          "author": "author 2",
          "description": "description 2",
          "id": "FROZEN-ID",
          "name": "name 2",
        },
      },
    ],
  },
}
`;

exports[`Book queries should query a connection of books with category filter 1`] = `
Object {
  "books": Object {
    "count": 5,
    "edges": Array [
      Object {
        "node": Object {
          "author": "author 5",
          "description": "description 5",
          "id": "FROZEN-ID",
          "name": "name 5",
        },
      },
      Object {
        "node": Object {
          "author": "author 4",
          "description": "description 4",
          "id": "FROZEN-ID",
          "name": "name 4",
        },
      },
      Object {
        "node": Object {
          "author": "author 3",
          "description": "description 3",
          "id": "FROZEN-ID",
          "name": "name 3",
        },
      },
      Object {
        "node": Object {
          "author": "author 2",
          "description": "description 2",
          "id": "FROZEN-ID",
          "name": "name 2",
        },
      },
      Object {
        "node": Object {
          "author": "author 1",
          "description": "description 1",
          "id": "FROZEN-ID",
          "name": "name 1",
        },
      },
    ],
  },
}
`;

exports[`Book queries should query a connection of books with orderBy filter 1`] = `
Object {
  "books": Object {
    "count": 3,
    "edges": Array [
      Object {
        "node": Object {
          "author": "author 1",
          "description": "description 1",
          "id": "FROZEN-ID",
          "name": "name 1",
        },
      },
      Object {
        "node": Object {
          "author": "author 2",
          "description": "description 2",
          "id": "FROZEN-ID",
          "name": "name 2",
        },
      },
      Object {
        "node": Object {
          "author": "author 3",
          "description": "description 3",
          "id": "FROZEN-ID",
          "name": "name 3",
        },
      },
    ],
  },
}
`;

exports[`Book queries should query a connection of books with search filter 1`] = `
Object {
  "books": Object {
    "count": 3,
    "edges": Array [
      Object {
        "node": Object {
          "author": "author 3",
          "description": "TEST 2",
          "id": "FROZEN-ID",
          "name": "name 3",
        },
      },
      Object {
        "node": Object {
          "author": "TEST 3",
          "description": "description 2",
          "id": "FROZEN-ID",
          "name": "name 2",
        },
      },
      Object {
        "node": Object {
          "author": "author 1",
          "description": "description 1",
          "id": "FROZEN-ID",
          "name": "TEST 1",
        },
      },
    ],
  },
}
`;
